common/cluster/metadata_test.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	enumspb "go.temporal.io/api/enums/v1"
+	persistencespb "go.temporal.io/server/api/persistence/v1"
+	"go.temporal.io/server/common/clock"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/common/searchattribute"
+)
+
+type (
+	metadataSuite struct {
+		suite.Suite
+		*require.Assertions
+
+		controller *gomock.Controller
+
+		logger                     log.Logger
+		timeSource                 *clock.EventTimeSource
+		mockClusterMetadataManager *persistence.MockClusterMetadataManager
+		metadata                   Metadata
+	}
+)
+
+func TestNamespaceCacheSuite(t *testing.T) {
+	suite.Run(t, &metadataSuite{})
+}
+
+func (s *metadataSuite) SetupSuite() {
+}
+
+func (s *metadataSuite) TearDownSuite() {
+
+}
+
+func (s *metadataSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+
+	s.controller = gomock.NewController(s.T())
+
+	s.logger = log.NewDefaultLogger()
+	s.timeSource = clock.NewEventTimeSource()
+	s.mockClusterMetadataManager = persistence.NewMockClusterMetadataManager(s.controller)
+	s.metadata = NewTestClusterMetadata(NewTestClusterMetadataConfig(false, true), s.mockClusterMetadataManager, s.timeSource)
+}
+
+func (s *metadataSuite) TearDownTest() {
+	s.controller.Finish()
+}
+
+func (s *metadataSuite) TestGetSearchAttributesCache() {
+	s.timeSource.Update(time.Date(2020, 8, 22, 1, 0, 0, 0, time.UTC))
+	// Initial call
+	s.mockClusterMetadataManager.EXPECT().GetClusterMetadata().Return(&persistence.GetClusterMetadataResponse{
+		ClusterMetadata: persistencespb.ClusterMetadata{
+			IndexSearchAttributes: map[string]*persistencespb.IndexSearchAttributes{
+				"index-name": {
+					SearchAttributes: map[string]enumspb.IndexedValueType{
+						"OrderId": enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+					}}},
+		},
+		Version: 1,
+	}, nil)
+	// Second call, no changes in DB (version is the same)
+	s.mockClusterMetadataManager.EXPECT().GetClusterMetadata().Return(&persistence.GetClusterMetadataResponse{
+		ClusterMetadata: persistencespb.ClusterMetadata{
+			IndexSearchAttributes: map[string]*persistencespb.IndexSearchAttributes{
+				"index-name": {
+					SearchAttributes: map[string]enumspb.IndexedValueType{
+						"OrderId": enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+					}}},
+		},
+		Version: 1,
+	}, nil)
+	// Third call, DB changed
+	s.mockClusterMetadataManager.EXPECT().GetClusterMetadata().Return(&persistence.GetClusterMetadataResponse{
+		ClusterMetadata: persistencespb.ClusterMetadata{
+			IndexSearchAttributes: map[string]*persistencespb.IndexSearchAttributes{
+				"index-name": {
+					SearchAttributes: map[string]enumspb.IndexedValueType{
+						"OrderId": enumspb.INDEXED_VALUE_TYPE_KEYWORD,
+					}}},
+		},
+		Version: 2,
+	}, nil)
+
+	wg := sync.WaitGroup{}
+	wg.Add(10)
+	for goroutine := 0; goroutine < 10; goroutine++ {
+		go func(goroutine int) {
+			for i := 1; i < 1500; i++ {
```

Comment:
This tests concurrent read/writes and makes sure that DB read is called only expected number of times.